nemo/collections/nlp/models/intent_slot_classification/multi_label_intent_slot_classification_model.py

```diff
@@ -0,0 +1,444 @@
+                probabilities.append(torch.sigmoid(intent_logits).detach().cpu().numpy())
+
+            probabilities = np.concatenate(probabilities)
+
+        finally:
+            # set mode back to its original value
+            self.train(mode=mode)
+
+        return probabilities
+
+    def optimize_threshold(self, test_ds: str, file_name: str) -> None:
+        """
+        Set the optimal threshold of the model from performance on validation set. This threshold is used to round the 
+        logits to 0 or 1. 
+
+        Args:
+            test_ds: location of test dataset
+            file_name: name of input file to retrieve validation set
+
+        Returns:
+            None
+        """
+
+        input_file = f"{self.data_dir}/{file_name}.tsv"
+
+        with open(input_file, "r") as f:
+            input_lines = f.readlines()[1:]  # Skipping headers at index 0
+
+        dataset = list(input_lines)
+
+        metrics_labels, sentences = [], []
+
+        for input_line in dataset:
+            sentence = input_line.strip().split("\t")[0]
+            sentences.append(sentence)
+            parts = input_line.strip().split("\t")[1:][0]
+            parts = list(map(int, parts.split(",")))
+            parts = [1 if label in parts else 0 for label in range(len(self.cfg.data_desc.intent_labels))]
+            metrics_labels.append(parts)
+
+        # Retrieve class probabilities for each sentence
+        intent_probabilities = self.prediction_probabilities(sentences, test_ds)
+
+        metrics_dict = {}
+        # Find optimal logits rounding threshold for intents
+        for i in np.arange(0.5, 0.96, 0.01):
+            predictions = (intent_probabilities >= i).tolist()
+            precision = precision_score(metrics_labels, predictions, average='micro')
+            recall = recall_score(metrics_labels, predictions, average='micro')
+            f1 = f1_score(metrics_labels, predictions, average='micro')
+            metrics_dict[i] = [precision, recall, f1]
+
+        max_precision = max(metrics_dict, key=lambda x: metrics_dict[x][0])
+        max_recall = max(metrics_dict, key=lambda x: metrics_dict[x][1])
+        max_f1_score = max(metrics_dict, key=lambda x: metrics_dict[x][2])
+
+        logging.info(
+            f'Best Threshold for F1-Score: {max_f1_score}, [Precision, Recall, F1-Score]: {metrics_dict[max_f1_score]}'
+        )
+        logging.info(
+            f'Best Threshold for Precision: {max_precision}, [Precision, Recall, F1-Score]: {metrics_dict[max_precision]}'
+        )
+        logging.info(
+            f'Best Threshold for Recall: {max_recall}, [Precision, Recall, F1-Score]: {metrics_dict[max_recall]}'
+        )
+
+        if metrics_dict[max_f1_score][2] > self.max_f1:
+            self.max_f1 = metrics_dict[max_f1_score][2]
+
+            logging.info(f'Setting Threshold to: {max_f1_score}')
+
+            self.threshold = max_f1_score
+
+    def predict_from_examples(
```

Comment:
I don't think you can set self.threshold as a default argument, so I set it as 0.5 at the moment.

Reply:
Ah yeah, you're right of course. The usual workaround for this situation is to set the default as None and then replace it with the object attribute inside the method. I think this will be better than using 0.5, which might be unexpected if a user has just gone through the threshold finding and thinks the threshold is now going to be used.